File: 383-ransom-note/383-ransom-note.go
func canConstruct(ransomNote string, magazine string) bool {
    magCounts := make(map[rune]int)
    
    for _, char := range magazine {
        magCounts[char] += 1
    }
    
    for _, char := range ransomNote {
        if count, exists := magCounts[char]; !exists || count == 0 {
            return false
        }
        magCounts[char] -= 1
    }
    
    return true
}

File: 383-ransom-note/383-ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
  const magCounts = {};
  
  for (const char of magazine) {
    magCounts[char] = (magCounts[char] ?? 0) + 1;
  }
  
  for (const char of ransomNote) {
    if (!magCounts.hasOwnProperty(char) || magCounts[char] === 0) {
      return false;
    }
    magCounts[char] -= 1;
  }
  
  return true;
};